editor/editor/mode/visual.js

```diff
@@ -1,34 +1,105 @@
 /**
+ * External dependencies
+ */
+import classNames from 'classnames';
+
+/**
  * Internal dependencies
  */
 import InserterButton from '../../inserter/button';
 
-function Blocks( { blocks, onChange } ) {
-	const onChangeBlock = ( index ) => ( changes ) => {
-		const newBlock = {
-			...blocks[ index ],
-			changes
+class Blocks extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.resetBlockState = this.resetBlockState.bind( this );
+		this.state = {
+			blocks: [],
+			selected: null,
+			hovered: null,
+		};
+	}
+
+	componentWillMount() {
+		this.setState( { blocks: this.props.blocks } );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.blocks !== this.state.blocks ) {
+			this.setState( { blocks: nextProps.blocks } );
+		}
+	}
+
+	onChangeBlock() {
+		return ( index ) => ( changes ) => {
+			const { blocks } = this.state;
+			const newBlock = {
+				...blocks[ index ],
+				attributes: {
+					...blocks[ index ].attributes,
+					...changes
+				}
+			};
+
+			const newBlocks = [
+				...blocks.slice( 0, index ),
+				newBlock,
+				...blocks.slice( index + 1 )
+			];
+
+			this.setState( { blocks: newBlocks } );
```

Comment:
Can we avoid the mix of state and props blocks by forcing the component to be used as a controlled component? i.e. all blocks changes go through `this.props.onChange` and we assume that `this.props.blocks` is the canonical set of blocks.